### PR TITLE
device-mapper udev sync

### PR DESCRIPTION
**Pull request by vbatts**
sync devicemapper and udev.
If sync is not supported, print a warning that unexpected behavior may occur.
relates to #4036 
ping @unclejack

### CODE REVIEW

File: daemon/graphdriver/devmapper/deviceset.go [hunk 1]
```diff
@@ -947,6 +947,12 @@ func (devices *DeviceSet) initDevmapper(doInit bool) error {
 		return graphdriver.ErrNotSupported
 	}
 
+	// https://github.com/docker/docker/issues/4036
+	if supported := devicemapper.UdevSetSyncSupport(true); !supported {
+		log.Warnf("Warning, Udev sync is not supported. Behavior may be unpredictable")
```
Comment (unclejack) [inline daemon/graphdriver/devmapper/deviceset.go]: @vbatts I'd change `Warning, Udev sync is not supported. Behavior may be unpredictable` to something along the lines of `WARNING: Udev sync is not supported. This will lead to unexpected behavior, data loss and errors`.
Perhaps we should also print the kind of errors (like the ones seen on #4036) one would see when printing this warning? This should make it easier to set expectations, rather than letting users ignore it.

Comment (vbatts) [inline daemon/graphdriver/devmapper/deviceset.go]: what about just a link to #4036?

Comment (unclejack) [inline daemon/graphdriver/devmapper/deviceset.go]: I've thought about that, but #4036 seems to have become a catchall for devicemapper related issues. It might end up being very confusing when trying to figure out exactly what kind of errors could be attributed to this and what could be caused by something else.

**Comment by vbatts** [comment]
@unclejack I just updated the warning

**Comment by icecrime** [comment]
LGTM, and also interested to see this setting's value in `docker info` output!

**Comment by vbatts** [comment]
@icecrime oh? well that is an idea too, and easy enough to do. Though I would not print a warning then, only in the daemon log.

**Comment by icecrime** [comment]
It doesn't necessarily have to be in this PR. It's just that I understand from reading the patch that udev sync really matters, so if we see issues filed for device mapper problems it would probably make sense to know the value of this setting for investigation, right?

**Comment by vbatts** [comment]
@icecrime done. PTAL.

**Comment by icecrime** [comment]
Perfect thanks! LGTM

**Comment by unclejack** [comment]
I've tested this on CentOS 7 with dynbinary and it's all working just fine.

This can be merged after the drone build is done.

LGTM

**Comment by rhvgoyal** [comment]
I am looking at libdm code.

dm_udev_set_sync_support() does not do anything if libdm was built without UDEV_SYNC_SUPPORT. If udev sync support was enabled, dm_udev_set_sync_support() sets
variable _sync_with_udev. And default value of _sync_with_udev is 1.

IOW, if libdm is built with udev sync support, it is enabled by default. If it is not built with udev sync support then one can not enable it at run time.

So calling dm_udev_set_sync_support() from docker should not help in anyway. Am I missing something?

**Comment by vbatts** [comment]
@rhvgoyal while `_sync_with_udev = 1` by default, the call to `dm_udev_set_sync_support` performs some one-time checks. Granted this is largely a no-op, because if it's compiled, it's `1`, and if it's not compiled it's `0`. The bigger functionality is calling of `dm_udev_get_sync_support`. Which does a couple of checks for support. 
Since we do this on init of the driver, we can perform the checks and ensure the sync support is reported in the log and `docker info`

**Comment by rhvgoyal** [comment]
ok, so this commit will just introduce warnings if udev_sync is not supported (when docker binary is built static) and make debugging little easier. But it should not fix udev races.

So we still have the problem that how to address the issue of libdm and udev races with statically built docker.

**Comment by vbatts** [comment]
@rhvgoyal that is a different story. static docker, means static libdm, which _would_ require libudev, which will not happen. 1) systemd/udev does not even allow static linking; 2) you would not want to have varying versions of udev (static in docker and whatever version on the host) because the API is not guaranteed across versions.

For static docker, you either get libdm vs. udev race or should disable the device-mapper driver entirely. Which in that case, sync_support will return 0/false.

Basically, we should get to the point of sanity checking for drivers, and if `dm_udev_get_sync_support == 0` then bail on using the devicemapper driver.

**Comment by davedoesdev** [comment]
I'm not running udev at all - can I ignore this warning?

**Comment by vbatts** [comment]
David, interesting that your OS is not running udev. Can you provide more
info on your setup so that I may recreate it for myself?

**Comment by vbatts** [comment]
Also, if udev is not present at all then `devicemapper` (libdm) will
fallback to managing the devices itself, and you should be able to ignore
this warning.

Though in docker 1.7 you will need to add an additional flag to proceed
past the udev sync check.

**Comment by davedoesdev** [comment]
Thanks - that'll be `--storage-opt dm.override_udev_sync_check=true`, right? That's not going to make it for 1.6?

I'm running my a Linux OS I built myself: https://github.com/davedoesdev/heddle

**Comment by vbatts** [comment]
That's the right flag. Hopefully you'll trial the different drivers you can
support. Also you should ensure you are running devicemapper not on
loopfile (which is the default), but actually setting up block devices.

**Comment by davedoesdev** [comment]
Thanks. I also support btrfs which I'm finding to be a great filesystem. I'll probably default to btrfs soon and leave the devicemapper backend on loopfile.
